oligo_designer_toolsuite/oligo_selection/_heuristic_selection_methods.py

```diff
@@ -2,79 +2,392 @@
 # imports
 ############################################
 
+from abc import abstractmethod
+from typing import List
+
+import networkx as nx
 import numpy as np
 import pandas as pd
-import networkx as nx
 from scipy.sparse import csr_matrix
 
+from oligo_designer_toolsuite.oligo_efficiency_filter import SetScoringBase
+
 ############################################
 # Heuristic Selection Methods
 ############################################
 
 
-def heuristic_selection_independent_set(
-    oligoset_init: list,
-    oligos_scores: pd.Series,
-    overlapping_matrix: csr_matrix,
-    overlapping_matrix_ids: list,
-    n_oligo: int,
-    ascending: bool,
-    n_trials: int = 100,
-):
+class OligoSelectionPolicy:
     """
-    This method empirically finds an optimal set of non-overlapping oligos based on their scores. It iteratively
-    selects the best non-overlapping oligos to form a set, evaluates their scores, and discards oligos unlikely
-    to produce a better set than the best found set.
-
-    :param oligos_scores: Series containing scores of oligos.
-    :type oligos_scores: pd.Series
-    :param overlapping_matrix: Dataframe indicating overlap between oligos.
-    :type overlapping_matrix: pd.DataFrame
-    :param n_oligo: The number of oligos to select in each set.
-    :type n_oligo: int
-    :param ascending: Determines if scores should be sorted in ascending order, defaults to True.
+    Abstract base class for defining oligo selection policies.
+
+    :param set_size_opt: Optimal size for each oligo set.
+    :type set_size_opt: int
+    :param set_size_min: Minimum size for each oligo set.
+    :type set_size_min: int
+    :param n_sets: Number of oligo sets to generate.
+    :type n_sets: int
+    :param ascending: Flag indicating whether to sort scores in ascending order (depends on the score, set to True if lower scores are better).
     :type ascending: bool
-    :param n_trials: Number of top scoring oligos to consider for set formation, defaults to 100.
-    :type n_trials: int
-    :return: Updated scores, and the best set found.
-    :rtype: tuple(dict, pd.Series, pd.Series)
+    :param set_scoring: Set scoring class.
+    :type set_scoring: SetScoringBase
     """
+    def __init__(
+        self, set_size_opt: int, set_size_min: int, n_sets: int, ascending: bool, set_scoring: SetScoringBase
+    ) -> None:
+        self.opt_oligoset_size = set_size_opt
+        self.min_oligoset_size = set_size_min
+        self.n_sets = n_sets
```

Comment:
this should be parametrized from the"OligosetGeneratorIndependentSet" class -> you have those parameters in the init but you don't use them in the class.

Reply:
I agree there's a problem, but I'd say actually remove them from `OligosetGeneratorIndependentSet` because they are not used there. Otherwise, I'd have to supply them through the apply function, no?

Reply:
just keep thge parameters in the selection policy